File: native/ios/VocaKit/Sources/VocaSession/Cloud/CloudSessionViewModel.swift
import Foundation
import VocaKit

// MARK: - Cloud session (CONTRACT §3, mirrors JSX `FaceToFaceLiveScreen`)
//
// Two WebRTC legs against the backend, push-to-talk gated server side, a 1s
// poll for turn events, a 15s metering watermark and a budget cut-off. Every
// timer goes through `SessionScheduler` and every side effect through the
// injected protocols, so the whole state machine runs under XCTest with fakes.

@MainActor
public final class CloudSessionViewModel: ObservableObject {
    @Published public private(set) var state: LiveState

    public static let turnFallbackSeconds: TimeInterval = 36
    public static let usageReportEvery = 15
    public static let pollInterval: TimeInterval = 1
    public static let topic = "Translation Session"

    // Dependencies
    private let config: SessionConfig
    private let api: VocaAPI
    private let history: SessionHistoryStore
    private let legFactory: PeerLegFactory
    private let scheduler: SessionScheduler
    private let microphone: MicrophonePermission

    // Session
    private var sessionId: String?
    private var legs: [Side: PeerLeg] = [:]
    private var pcIds: [Side: String] = [:]
    /// Leg A's pc_id, kept past teardown: the transcript is re-projected after
    /// the session ends and turn attribution still has to work.
    private var speakerPcIdA: String?
    private var transcript: [SessionTurn] = []
    private var elapsedSeconds = 0
    private var connectionNote = "Connecting…"
    private var phase: LivePhase = .connecting
    private var tornDown = false
    /// `end()` is called from a button; two taps must not persist twice.
    private var ending = false

    // Push-to-talk
    private var pressing: [Side: Bool] = [.a: false, .b: false]
    private var turnState: [Side: TurnState] = [.a: .ready, .b: .ready]
    private var pressToken: [Side: Int] = [.a: 0, .b: 0]
    private var gateQueue: [Side: Task<PTTResult, Error>] = [:]
    private var turnTimeouts: [Side: SchedulerToken] = [:]

    // Timers
    private var timer: SchedulerToken?
    private var poller: SchedulerToken?
    private var pollInFlight = false
    private let tasks = TaskTracker()

    public init(config: SessionConfig, api: VocaAPI, history: SessionHistoryStore,
                legFactory: PeerLegFactory? = nil, scheduler: SessionScheduler = TaskScheduler(),
                microphone: MicrophonePermission = AVMicrophonePermission()) {
        self.config = config
        self.api = api
        self.history = history
        self.legFactory = legFactory ?? WebRTCLegFactory()
        self.scheduler = scheduler
        self.microphone = microphone
        self.state = LiveState(
            sideA: SideState(name: config.nameA, langCode: config.langA, langLabel: SessionLanguages.label(config.langA)),
            sideB: SideState(name: config.nameB, langCode: config.langB, langLabel: SessionLanguages.label(config.langB))
        )
        publish()
    }

    // MARK: Lifecycle

    /// §3 steps 1–5. Fire-and-forget; progress is reported through `state`.
    public func start() {
        tasks.track { [weak self] in await self?.setupSession() }
    }

    /// End pressed: persist, report usage, hang up, done.
    public func end() {
        guard !tornDown, !ending else { return }
        ending = true
        tasks.track { [weak self] in
            guard let self else { return }
            await self.persistSession(status: "ended")
            await self.reportElapsed()
            self.teardown()
            self.phase = .ended(.user)
            self.publish()
        }
    }

    /// Await every in-flight task (setup, gates, polls). Tests use this so
    /// assertions run after the fire-and-forget work has settled.
    public func settle() async {
        await tasks.settle()
    }

    private func setupSession() async {
        do {
            // 1. Mic
            guard await microphone.request() else {
                fail(.mic)
                return
            }
            if tornDown { return }

            // 2. ICE servers — an empty list fails the session.
            let iceServers = try await api.iceServers()
            guard !iceServers.isEmpty else { throw CloudSessionError.noIceServers }
            if tornDown { return }

            // 3. Session
            let sessionId = try await api.createSession(CreateSessionRequest(
                callerName: config.nameA, callerLanguage: config.langA,
                topic: CloudSessionViewModel.topic, clientId: config.clientId
            ))
            if tornDown { return }
            self.sessionId = sessionId
            tasks.track { [weak self] in await self?.persistSession(status: "active") }

            // 4. Two relay-only peer connections, mic cloned per leg (enabled
            //    for negotiation — see §3 step 3).
            let legA = try legFactory.makeLeg(iceServers: iceServers)
            let legB = try legFactory.makeLeg(iceServers: iceServers)
            legs[.a] = legA
            legs[.b] = legB

            // 5. Connection state: leg A is the one that starts the clocks.
            legA.onConnectionState = { [weak self] connectionState in
                self?.handleConnectionState(connectionState)
            }
            legB.onConnectionState = { [weak self] connectionState in
                guard let self, connectionState == .failed || connectionState == .disconnected else { return }
                self.handleConnectionState(connectionState)
            }

            // 6. Offers, one per leg, then exchange with the server in order.
            let offerA = try await legA.createOffer()
            let offerB = try await legB.createOffer()
            if tornDown { return }

            let answerA = try await api.offer(OfferRequest(
                sessionId: sessionId, language: config.langA, name: config.nameA, sdp: offerA.sdp, type: offerA.type
            ))
            if tornDown { return }
            pcIds[.a] = answerA.pcId
            speakerPcIdA = answerA.pcId
            try await legA.setRemoteAnswer(sdp: answerA.sdp, type: answerA.type)
            if tornDown { return }

            let answerB = try await api.offer(OfferRequest(
                sessionId: sessionId, language: config.langB, name: config.nameB, sdp: offerB.sdp, type: offerB.type
            ))
            if tornDown { return }
            pcIds[.b] = answerB.pcId
            try await legB.setRemoteAnswer(sdp: answerB.sdp, type: answerB.type)
            if tornDown { return }

            // RTP is established; push-to-talk owns the tracks from here on.
            legA.setTrackEnabled(false)
            legB.setTrackEnabled(false)
        } catch {
            if tornDown { return }
            fail(.network)
        }
    }

    private func handleConnectionState(_ connectionState: PeerConnectionState) {
        guard !tornDown else { return }
        switch connectionState {
        case .connected:
            guard phase == .connecting else { return }
            phase = .live
            connectionNote = "Connected"
            startTimer()
            startPoll()
            publish()
        case .failed, .disconnected:
            connectionNote = "Connection lost"
            fail(.network)
        default:
            break
        }
    }

    private func fail(_ error: SessionError) {
        guard !tornDown else { return }
        if error == .network, connectionNote == "Connecting…" { connectionNote = "Connection failed" }
        teardown()
        phase = .error(error)
        publish()
    }

    // MARK: Timer + metering (§3 step 9)

    private func startTimer() {
        guard timer == nil else { return }
        timer = scheduler.repeating(every: 1) { [weak self] in self?.tick() }
    }

    private func tick() {
        guard !tornDown else { return }
        elapsedSeconds += 1
        let spent = elapsedSeconds

        if spent % CloudSessionViewModel.usageReportEvery == 0 {
            tasks.track { [weak self] in await self?.reportElapsed() }
        }

        // Out of credit: end the call rather than let it run on unbilled. The
        // transcript is persisted first so the user keeps what was said.
        if TimeInterval(spent) >= config.budgetSeconds {
            tasks.track { [weak self] in
                guard let self else { return }
                await self.reportElapsed()
                await self.persistSession(status: "ended")
                self.teardown()
                self.phase = .ended(.exhausted)
                self.publish()
            }
            return
        }
        publish()
    }

    private func reportElapsed() async {
        guard let sessionId else { return }
        _ = await api.consume(subject: config.clientId, sessionId: sessionId, seconds: elapsedSeconds)
    }

    // MARK: Poll (§3 step 8)

    private func startPoll() {
        guard poller == nil else { return }
        poller = scheduler.repeating(every: CloudSessionViewModel.pollInterval) { [weak self] in
            self?.tasks.track { [weak self] in await self?.pollOnce() }
        }
    }

    private func pollOnce() async {
        guard !tornDown, !pollInFlight, let sessionId else { return }
        pollInFlight = true
        defer { pollInFlight = false }

        let response: PollResponse
        do {
            response = try await api.poll(sessionId: sessionId)
        } catch {
            return
        }
        guard !tornDown else { return }

        var sawTurn = false
        for event in response.events {
            switch event.type {
            case "turn":
                sawTurn = true
                let turn = SessionTurn(
                    speaker: event.speaker, speakerName: event.speakerName,
                    originalLang: event.originalLang ?? "", original: event.original ?? "",
                    translated: event.translated ?? ""
                )
                transcript.append(turn)
                markTurnReady(spokenBy(turn))
            case "turn_failed":
                // The server gave up on this turn; release the speaker now
                // instead of waiting for the 36s client fallback.
                markTurnReady(event.speaker == speakerPcIdA ? .a : .b)
            default:
                break // status and unknown types are ignored
            }
        }
        if sawTurn {
            tasks.track { [weak self] in await self?.persistSession(status: "active") }
        }
        publish()

        if response.closed {
            await persistSession(status: "ended")
            teardown()
            phase = .ended(.serverClosed)
            publish()
        }
    }

    /// Attribute by `speaker == pc_id`; fall back to the original language.
    private func spokenBy(_ turn: SessionTurn) -> Side {
        if let speaker = turn.speaker, !speaker.isEmpty {
            return speaker == speakerPcIdA ? .a : .b
        }
        return turn.originalLang == config.langA ? .a : .b
    }

    // MARK: Push-to-talk (§3 step 7)

    public func hold(side: Side) {
        guard phase == .live, turnState[side] == .ready, pressing[side.other] == false else { return }
        let token = (pressToken[side] ?? 0) + 1
        pressToken[side] = token

        legs[side.other]?.setTrackEnabled(false)
        pressing[side.other] = false
        pressing[side] = true
        turnState[side] = .recording
        publish()

        let gate = queueGate(side, .hold)
        tasks.track { [weak self] in
            guard let self else { return }
            do {
                _ = try await gate.value
                // Only unmute if this press is still the live one.
                if self.pressToken[side] == token, self.pressing[side] == true {
                    self.legs[side]?.setTrackEnabled(true)
                }
            } catch {
                if self.pressToken[side] == token, self.pressing[side] == true {
                    self.pressing[side] = false
                    self.turnState[side] = .ready
                    self.publish()
                }
            }
        }
    }

    public func release(side: Side) {
        guard pressing[side] == true else { return }
        legs[side]?.setTrackEnabled(false)
        pressing[side] = false
        markTurnTranslating(side)
        publish()

        let gate = queueGate(side, .release)
        tasks.track { [weak self] in
            guard let self else { return }
            do {
                let result = try await gate.value
                // Nothing captured → nothing to wait for.
                if (result.flushedBytes ?? 0) == 0 { self.markTurnReady(side) }
            } catch {
                self.markTurnReady(side)
            }
            self.publish()
        }
    }

    /// Per-side FIFO so a quick hold/release never reaches the server out of order.
    private func queueGate(_ side: Side, _ action: PTTAction) -> Task<PTTResult, Error> {
        let previous = gateQueue[side]
        let pcId = pcIds[side]
        let api = self.api
        let task = Task<PTTResult, Error> { @MainActor in
            _ = await previous?.result
            guard let pcId else { throw CloudSessionError.notConnected }
            return try await api.ptt(pcId: pcId, action: action)
        }
        gateQueue[side] = task
        return task
    }

    private func markTurnTranslating(_ side: Side) {
        clearTurnTimeout(side)
        turnState[side] = .translating
        // If an upstream translation fails without producing a turn event,
        // don't strand this speaker forever (same window as the backend).
        turnTimeouts[side] = scheduler.after(CloudSessionViewModel.turnFallbackSeconds) { [weak self] in
            self?.markTurnReady(side)
            self?.publish()
        }
    }

    private func markTurnReady(_ side: Side) {
        clearTurnTimeout(side)
        guard !tornDown else { return }
        turnState[side] = .ready
    }

    private func clearTurnTimeout(_ side: Side) {
        turnTimeouts[side]?.cancel()
        turnTimeouts[side] = nil
    }

    // MARK: Persistence + teardown (§3 step 10)

    private func persistSession(status: String) async {
        guard let sessionId else { return }
        let record = SessionRecord(
            sessionId: sessionId,
            callerName: config.nameA,
            topic: CloudSessionViewModel.topic,
            languageA: config.langA,
            languageB: config.langB,
            participantA: config.nameA,
            participantB: config.nameB,
            status: status,
            durationSeconds: elapsedSeconds,
            transcriptJSON: SessionTurn.encodeTranscript(transcript)
        )
        await history.save(record)
    }

    private func teardown() {
        guard !tornDown else { return }
        tornDown = true
        timer?.cancel(); timer = nil
        poller?.cancel(); poller = nil
        clearTurnTimeout(.a)
        clearTurnTimeout(.b)

        for side in Side.allCases {
            if let pcId = pcIds[side] {
                let api = self.api
                tasks.track { await api.hangup(pcId: pcId) }
            }
            legs[side]?.close()
        }
        pcIds.removeAll()
        legs.removeAll()

        pressing = [.a: false, .b: false]
        turnState = [.a: .ready, .b: .ready]
    }

    // MARK: Projection → LiveState (JSX render block)

    private func publish() {
        let live = phase == .live
        let spokenA = transcript.map { spokenBy($0) == .a }
        let hasSpokenA = spokenA.contains(true)
        let hasSpokenB = spokenA.contains(false)

        func side(_ side: Side, name: String, otherName: String, hasSpoken: Bool) -> SideState {
            let pressingThis = pressing[side] == true
            let pressingOther = pressing[side.other] == true
            let translating = turnState[side] == .translating
            let status: String
            if pressingThis { status = "Recording" }
            else if translating { status = "Translating..." }
            else if pressingOther { status = "Listening to \(otherName)" }
            else if live { status = hasSpoken ? "Ready - speak again" : "Ready" }
            else { status = connectionNote }

            let turns = transcript.enumerated().map { index, turn in
                TurnView(id: index, original: turn.original, translated: turn.translated,
                         mine: spokenA[index] == (side == .a))
            }
            let code = side == .a ? config.langA : config.langB
            return SideState(
                name: name, langCode: code, langLabel: SessionLanguages.label(code),
                pressing: pressingThis,
                disabled: !live || translating || pressingOther,
                dimmed: translating,
                status: status,
                turns: turns,
                note: (phase == .connecting && transcript.isEmpty) ? connectionNote : nil
            )
        }

        state = LiveState(
            sideA: side(.a, name: config.nameA, otherName: config.nameB, hasSpoken: hasSpokenA),
            sideB: side(.b, name: config.nameB, otherName: config.nameA, hasSpoken: hasSpokenB),
            elapsed: TimeInterval(elapsedSeconds),
            phase: phase,
            badge: nil,
            notice: nil
        )
    }
}

// MARK: - Transcript JSON (matches the JSX event shape the history store already reads)

extension SessionTurn {
    private struct Wire: Codable {
        var speaker: String?
        var speaker_name: String?
        var original_lang: String
        var target_lang: String?
        var original: String
        var translated: String
        var side: String?
    }

    public static func encodeTranscript(_ turns: [SessionTurn], sides: [Side]? = nil) -> String {
        let wire = turns.enumerated().map { index, turn in
            Wire(speaker: turn.speaker, speaker_name: turn.speakerName, original_lang: turn.originalLang,
                 target_lang: turn.targetLang, original: turn.original, translated: turn.translated,
                 side: sides.flatMap { index < $0.count ? $0[index].rawValue : nil })
        }
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys]
        guard let data = try? encoder.encode(wire), let text = String(data: data, encoding: .utf8) else { return "[]" }
        return text
    }
}

File: native/android/session/src/main/kotlin/com/vocare/translate/session/cloud/CloudSessionViewModel.kt
package com.vocare.translate.session.cloud

import androidx.lifecycle.ViewModel
import com.vocare.translate.core.api.ApiException
import com.vocare.translate.core.api.VocaApi
import com.vocare.translate.core.history.SessionHistoryStore
import com.vocare.translate.core.model.CreateSessionRequest
import com.vocare.translate.core.model.Language
import com.vocare.translate.core.model.LiveState
import com.vocare.translate.core.model.OfferRequest
import com.vocare.translate.core.model.PollEvent
import com.vocare.translate.core.model.PttAction
import com.vocare.translate.core.model.SessionConfig
import com.vocare.translate.core.model.SessionError
import com.vocare.translate.core.model.SessionPhase
import com.vocare.translate.core.model.SessionRecord
import com.vocare.translate.core.model.Side
import com.vocare.translate.core.model.SideState
import com.vocare.translate.core.model.Turn
import com.vocare.translate.core.model.TurnView
import com.vocare.translate.session.EndReason
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/** How the host asks for `RECORD_AUDIO`; the module cannot show the prompt itself. */
fun interface MicPermission {
    suspend fun request(): Boolean

    companion object {
        val Granted = MicPermission { true }
    }
}

/**
 * Face-to-face cloud session: two TURN-only legs, server-side push-to-talk
 * gates, a 1s poll loop and idempotent metering. Implements CONTRACT §3 step by
 * step and mirrors JSX `FaceToFaceLiveScreen`.
 *
 * Everything runs on [dispatcher] (the main thread in the app, a test dispatcher
 * in tests). Time is only ever consumed through `delay`, so tests drive the 1s
 * tick, the 15s consume watermark and the 36s turn fallback with virtual time.
 */
class CloudSessionViewModel(
    private val config: SessionConfig,
    private val api: VocaApi,
    private val history: SessionHistoryStore,
    private val legFactory: PeerLegFactory,
    private val micPermission: MicPermission = MicPermission.Granted,
    dispatcher: CoroutineDispatcher = Dispatchers.Main.immediate,
) : ViewModel() {

    private enum class TurnState { READY, RECORDING, TRANSLATING }

    private class SideRuntime {
        var turnState = TurnState.READY
        var pressing = false

        /** JSX pointer id: identifies the press whose hold gate is in flight. */
        var activePress: Int? = null
        var pressSerial = 0
        var leg: PeerLeg? = null
        var pcId: String? = null

        /** Fair FIFO — serialises hold/release per side (JSX `gateQueue`). */
        val gate = Mutex()
        var fallback: Job? = null
    }

    private val scope = CoroutineScope(SupervisorJob() + dispatcher)

    /** Hangups and the final consume must outlive the session scope (JSX `fireAndForgetHangup`). */
    private val detachedScope = CoroutineScope(SupervisorJob() + dispatcher)

    private val sides = mapOf(Side.A to SideRuntime(), Side.B to SideRuntime())
    private val transcript = mutableListOf<Turn>()
    private var sessionId: String? = null
    private var elapsed = 0
    private var phase: SessionPhase = SessionPhase.Connecting
    private var connState = "Connecting…"
    private var timerJob: Job? = null
    private var pollJob: Job? = null
    private var setupJob: Job? = null
    private var tornDown = false

    private val _state = MutableStateFlow(render())
    val state: StateFlow<LiveState> = _state.asStateFlow()

    private val _endReason = MutableStateFlow<EndReason?>(null)

    /** Set together with `SessionPhase.Ended`; the host routes `EXHAUSTED` to the paywall. */
    val endReason: StateFlow<EndReason?> = _endReason.asStateFlow()

    val currentSessionId: String? get() = sessionId

    // ── §3.1–3.5 setup ───────────────────────────────────────────────────────

    fun start() {
        if (setupJob != null) return
        setupJob = scope.launch { setup() }
    }

    private suspend fun setup() {
        try {
            if (!micPermission.request()) {
                fail(SessionError.MIC, "Microphone access denied")
                return
            }
            val iceServers = api.iceServers()
            if (iceServers.isEmpty()) throw ApiException.NoIceServers()

            val id = api.createSession(
                CreateSessionRequest(
                    callerName = config.nameA,
                    callerLanguage = config.langA,
                    topic = TOPIC,
                    clientId = config.clientId,
                ),
            )
            sessionId = id
            persist(STATUS_ACTIVE)

            val legA = legFactory.create(iceServers) { onLegState(Side.A, it) }
            val legB = legFactory.create(iceServers) { onLegState(Side.B, it) }
            sides.getValue(Side.A).leg = legA
            sides.getValue(Side.B).leg = legB
            if (tornDown) return

            // Both clones are enabled while negotiating (see PeerLeg.createOffer).
            val offerA = legA.createOffer()
            val offerB = legB.createOffer()
            if (tornDown) return

            connectLeg(Side.A, id, offerA)
            if (tornDown) return
            connectLeg(Side.B, id, offerB)
            if (tornDown) return

            // RTP is established; push-to-talk owns the tracks from here on.
            legA.setMicEnabled(false)
            legB.setMicEnabled(false)
        } catch (e: CancellationException) {
            throw e
        } catch (e: ApiException.PaymentRequired) {
            // 402 on /offer: metering refused the leg. Persist what exists and let the host
            // route to the `exhausted` paywall rather than a misleading network error.
            finish(EndReason.EXHAUSTED)
        } catch (e: Exception) {
            fail(SessionError.NETWORK, "Connection failed")
        }
    }

    private suspend fun connectLeg(side: Side, id: String, offer: LocalOffer) {
        val runtime = sides.getValue(side)
        val answer = api.offer(
            OfferRequest(
                sessionId = id,
                language = lang(side),
                name = name(side),
                sdp = offer.sdp,
                type = offer.type,
            ),
        )
        runtime.pcId = answer.pcId
        runtime.leg?.setAnswer(answer.sdp, answer.type)
    }

    /** §3.6: only leg A's state drives the screen, as in the JSX. Called from any thread. */
    private fun onLegState(side: Side, state: LegConnectionState) {
        if (side != Side.A) return
        scope.launch {
            if (tornDown || phase is SessionPhase.Ended) return@launch
            when (state) {
                LegConnectionState.CONNECTED -> {
                    if (phase is SessionPhase.Live) return@launch
                    phase = SessionPhase.Live
                    connState = "Connected"
                    startTimer()
                    startPoll()
                    publish()
                }
                LegConnectionState.FAILED, LegConnectionState.DISCONNECTED -> {
                    fail(SessionError.NETWORK, "Connection lost")
                }
                else -> Unit
            }
        }
    }

    // ── §3.7 push-to-talk ────────────────────────────────────────────────────

    fun hold(side: Side) {
        val me = sides.getValue(side)
        val other = sides.getValue(side.other)
        if (phase !is SessionPhase.Live) return
        if (me.turnState != TurnState.READY || other.pressing) return

        other.leg?.setMicEnabled(false)
        other.pressing = false
        me.pressing = true
        me.turnState = TurnState.RECORDING
        val token = ++me.pressSerial
        me.activePress = token
        publish()

        scope.launch {
            me.gate.withLock {
                try {
                    api.ptt(requirePcId(me), PttAction.HOLD)
                    // Only unmute once the server-side gate is open — and only if this
                    // press is still the live one (JSX pointerId check).
                    if (me.activePress == token) me.leg?.setMicEnabled(true)
                } catch (e: CancellationException) {
                    throw e
                } catch (e: Exception) {
                    if (me.activePress == token) {
                        me.activePress = null
                        me.pressing = false
                        me.turnState = TurnState.READY
                        publish()
                    }
                }
            }
        }
    }

    fun release(side: Side) {
        val me = sides.getValue(side)
        if (me.activePress == null) return
        me.activePress = null
        me.leg?.setMicEnabled(false)
        me.pressing = false
        markTranslating(side)
        publish()

        scope.launch {
            me.gate.withLock {
                try {
                    val result = api.ptt(requirePcId(me), PttAction.RELEASE)
                    // Nothing captured → nothing will come back; don't wait for a turn.
                    if ((result.flushedBytes ?: 0L) <= 0L) markReady(side)
                } catch (e: CancellationException) {
                    throw e
                } catch (e: Exception) {
                    markReady(side)
                }
            }
        }
    }

    private fun requirePcId(runtime: SideRuntime): String =
        runtime.pcId ?: throw IllegalStateException("Translation participant is not connected")

    private fun markTranslating(side: Side) {
        val me = sides.getValue(side)
        me.fallback?.cancel()
        me.turnState = TurnState.TRANSLATING
        // If the upstream translation never produces a turn, don't strand this speaker.
        me.fallback = scope.launch {
            delay(TURN_FALLBACK_MS)
            markReady(side)
        }
    }

    private fun markReady(side: Side) {
        val me = sides.getValue(side)
        me.fallback?.cancel()
        me.fallback = null
        if (tornDown) return
        me.turnState = TurnState.READY
        publish()
    }

    // ── §3.6/§3.8 poll ───────────────────────────────────────────────────────

    private fun startPoll() {
        if (pollJob != null) return
        pollJob = scope.launch {
            while (isActive) {
                delay(POLL_INTERVAL_MS)
                val id = sessionId ?: continue
                val response = try {
                    api.poll(id)
                } catch (e: CancellationException) {
                    throw e
                } catch (e: Exception) {
                    continue
                }
                if (tornDown) return@launch
                var gotTurn = false
                for (event in response.events) {
                    when {
                        event.isTurn -> {
                            gotTurn = true
                            handleTurn(event)
                        }
                        event.isTurnFailed -> markReady(sideFor(event.speaker, event.originalLang))
                        else -> Unit // status, live and unknown types are ignored
                    }
                }
                if (gotTurn) persist(STATUS_ACTIVE)
                if (response.closed) {
                    finish(EndReason.SERVER_CLOSED)
                    return@launch
                }
            }
        }
    }

    /** §3.8 attribution: `speaker == pc_id_a`, falling back to `original_lang == langA`. */
    private fun sideFor(speaker: String?, originalLang: String?): Side {
        val pcIdA = sides.getValue(Side.A).pcId
        return if (speaker != null && pcIdA != null) {
            if (speaker == pcIdA) Side.A else Side.B
        } else {
            if (originalLang == config.langA) Side.A else Side.B
        }
    }

    private fun handleTurn(event: PollEvent) {
        transcript += event.toTurn()
        markReady(sideFor(event.speaker, event.originalLang))
    }

    // ── §3.9 timer + metering ────────────────────────────────────────────────

    private fun startTimer() {
        if (timerJob != null) return
        timerJob = scope.launch {
            while (isActive) {
                delay(TICK_MS)
                elapsed += 1
                if (elapsed % USAGE_REPORT_EVERY == 0) reportElapsed()
                if (elapsed >= config.budgetSeconds) {
                    // Out of credit: end rather than run on unbilled. Transcript is kept.
                    reportElapsed()
                    finish(EndReason.EXHAUSTED)
                    return@launch
                }
                publish()
            }
        }
    }

    /** Total elapsed seconds, never a delta — the server keeps a watermark. */
    private fun reportElapsed() {
        val id = sessionId ?: return
        val seconds = elapsed
        detachedScope.launch { api.consume(subject = config.clientId, sessionId = id, seconds = seconds) }
    }

    // ── §3.10 end / teardown ─────────────────────────────────────────────────

    /** User pressed End. */
    fun end() {
        if (phase is SessionPhase.Ended) return
        scope.launch {
            reportElapsed()
            finish(EndReason.USER)
        }
    }

    private suspend fun finish(reason: EndReason) {
        if (phase is SessionPhase.Ended) return
        persist(STATUS_ENDED)
        phase = SessionPhase.Ended
        _endReason.value = reason
        teardown()
        publish()
    }

    private fun fail(error: SessionError, message: String) {
        if (phase is SessionPhase.Ended) return
        connState = message
        phase = SessionPhase.Error(error)
        teardown()
        publish()
    }

    private fun teardown() {
        if (tornDown) return
        tornDown = true
        timerJob?.cancel(); timerJob = null
        pollJob?.cancel(); pollJob = null
        setupJob?.cancel()
        for ((_, runtime) in sides) {
            runtime.fallback?.cancel(); runtime.fallback = null
            runtime.activePress = null
            runtime.pressing = false
            runtime.turnState = TurnState.READY
            runtime.pcId?.let { pcId -> detachedScope.launch { api.hangup(pcId) } }
            runtime.pcId = null
            runtime.leg?.close()
            runtime.leg = null
        }
    }

    private suspend fun persist(status: String) {
        val id = sessionId ?: return
        history.save(
            SessionRecord(
                sessionId = id,
                callerName = config.nameA,
                topic = TOPIC,
                languageA = config.langA,
                languageB = config.langB,
                participantA = config.nameA,
                participantB = config.nameB,
                status = status,
                durationSeconds = elapsed,
                transcript = transcript.toList(),
            ),
        )
    }

    override fun onCleared() {
        if (phase !is SessionPhase.Ended && !tornDown) {
            // Leaving the screen without End: persist and hang up.
            detachedScope.launch { persist(STATUS_ENDED) }
            teardown()
        }
        scope.cancel()
    }

    // ── state projection (JSX render) ────────────────────────────────────────

    private fun publish() {
        _state.value = render()
    }

    private fun render(): LiveState = LiveState(
        sideA = renderSide(Side.A),
        sideB = renderSide(Side.B),
        elapsedSeconds = elapsed,
        phase = phase,
        badge = null,
    )

    private fun renderSide(side: Side): SideState {
        val me = sides.getValue(side)
        val other = sides.getValue(side.other)
        val live = phase is SessionPhase.Live
        val translating = me.turnState == TurnState.TRANSLATING
        val turns = transcript.mapIndexed { i, t ->
            TurnView(key = i, turn = t, mine = sideFor(t.speaker, t.originalLang) == side)
        }
        val hasSpoken = turns.any { it.mine }
        return SideState(
            name = name(side),
            langCode = lang(side),
            langLabel = Language.of(lang(side)).label,
            pressing = me.pressing,
            disabled = !live || translating || other.pressing,
            dimmed = translating,
            status = when {
                me.pressing -> "Recording"
                translating -> "Translating..."
                other.pressing -> "Listening to ${name(side.other)}"
                live -> if (hasSpoken) "Ready - speak again" else "Ready"
                else -> connState
            },
            turns = turns,
            note = if (phase is SessionPhase.Connecting && turns.isEmpty()) connState else null,
        )
    }

    private fun lang(side: Side) = if (side == Side.A) config.langA else config.langB
    private fun name(side: Side) = if (side == Side.A) config.nameA else config.nameB

    companion object {
        const val TOPIC = "Translation Session"
        const val STATUS_ACTIVE = "active"
        const val STATUS_ENDED = "ended"
        const val TICK_MS = 1_000L
        const val POLL_INTERVAL_MS = 1_000L
        const val USAGE_REPORT_EVERY = 15
        const val TURN_FALLBACK_MS = 36_000L
    }
}
